Mathlib/Order/Synonym.lean

```diff
@@ -0,0 +1,196 @@
+/-
+Copyright (c) 2020 Johan Commelin, Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Damiano Testa, Yaël Dillies
+-/
+import Mathlib.Logic.Equiv.Defs
+import Mathlib.Logic.Nontrivial
+import Mathlib.Order.Basic
+
+/-!
+# Type synonyms
+
+This file provides two type synonyms for order theory:
+* `OrderDual α`: Type synonym of `α` to equip it with the dual order (`a ≤ b` becomes `b ≤ a`).
+* `Lex α`: Type synonym of `α` to equip it with its lexicographic order. The precise meaning depends
+  on the type we take the lex of. Examples include `Prod`, `Sigma`, `List`, `Finset`.
+
+## Notation
+
+`αᵒᵈ` is notation for `OrderDual α`.
+
+The general rule for notation of `lex` types is to append `ₗ` to the usual notation.
+
+## Implementation notes
+
+One should not abuse definitional equality between `α` and `αᵒᵈ`/`Lex α`. Instead, explicit
+coercions should be inserted:
+* `OrderDual`: `OrderDual.toDual : α → αᵒᵈ` and `OrderDual.ofDual : αᵒᵈ → α`
+* `Lex`: `toLex : α → Lex α` and `ofLex : Lex α → α`.
+
+In fact, those are bundled as `Equiv`s to put goals in the right syntactic form for rewriting with
+the `Equiv` API (`⇑toLex a` where `⇑` is `coe_fn : (α ≃ Lex α) → α → Lex α`, instead of a bare
+`toLex a`).
```

Comment:
We'll need to update this comment to reflect that there is no longer a `coe_fn`!